File: tests/runtime/resources/resources.cs
using System.Diagnostics;
using ResourcesWorld;
using ResourcesWorld.wit.Imports;

namespace ResourcesWorld.wit.Exports
{
    public class ExportsExportsImpl : IExportsExports
    {
        public static IExportsExports.Z Add(IExportsExports.Z a, IExportsExports.Z b)
        {
            return new Z(((Z) a).val + ((Z) b).val);
        }

        public static void Consume(IExportsExports.X x)
        {
            x.Dispose();
        }
        
        public static void TestImports()
        {
            var y1 = new IImportsImports.Y(10);
            Debug.Assert(y1.GetA() == 10);
	    y1.SetA(20);
            Debug.Assert(y1.GetA() == 20);	    
	    var y2 = IImportsImports.Y.Add(y1, 20);
            Debug.Assert(y2.GetA() == 40);

	    var y3 = new IImportsImports.Y(1);
	    var y4 = new IImportsImports.Y(2);
            Debug.Assert(y3.GetA() == 1);
            Debug.Assert(y4.GetA() == 2);
	    y3.SetA(10);
	    y4.SetA(20);	    
            Debug.Assert(y3.GetA() == 10);
            Debug.Assert(y4.GetA() == 20);	    	    
	    var y5 = IImportsImports.Y.Add(y3, 20);
	    var y6 = IImportsImports.Y.Add(y4, 30);	    
            Debug.Assert(y5.GetA() == 30);
            Debug.Assert(y6.GetA() == 50);
        }

        public class X : IExportsExports.X, IExportsExports.IX {
            public int val;

            public X(int val) {
                this.val = val;
            }

            public void SetA(int val) {
                this.val = val;
            }

            public int GetA() {
                return val;
            }

            public static IExportsExports.X Add(IExportsExports.X a, int b) {
                var myA = (X) a;
                myA.SetA(myA.GetA() + b);
                return myA;
            }
        }
    
        public class Z : IExportsExports.Z, IExportsExports.IZ {
            private static uint numDropped = 0;
            
            public int val;

            public Z(int val) {
                this.val = val;
            }

            public int GetA() {
                return val;
            }

            public static uint NumDropped() {
                return numDropped + 1;
            }

            override protected void Dispose(bool disposing) {
		numDropped += 1;
                
                base.Dispose(disposing);
            }
        }

        public class KebabCase : IExportsExports.KebabCase, IExportsExports.IKebabCase {
            public uint val;
            
            public KebabCase(uint val) {
                this.val = val;
            }
            
            public uint GetA() {
                return val;
            }

            public static uint TakeOwned(IExportsExports.KebabCase a) {
                return ((KebabCase) a).val;
            }
        }
    }
}

File: tests/runtime/resources/resources.rs
include!(env!("BINDINGS"));

use exports::exports::{KebabCase, ZBorrow, X, Z};
use std::cell::RefCell;

pub struct Test {}

export!(Test);

pub struct ComponentX {
    val: RefCell<i32>,
}

pub struct ComponentZ {
    val: i32,
}

pub struct ComponentKebabCase {
    val: u32,
}

impl exports::exports::Guest for Test {
    type X = ComponentX;
    type Z = ComponentZ;
    type KebabCase = ComponentKebabCase;

    fn add(a: ZBorrow<'_>, b: ZBorrow<'_>) -> Z {
        let a = a.get::<ComponentZ>();
        let b = b.get::<ComponentZ>();
        Z::new(ComponentZ { val: a.val + b.val })
    }

    fn consume(x: exports::exports::X) {
        drop(x);
    }

    fn test_imports() -> Result<(), String> {
        use imports::*;
        let y = Y::new(10);
        assert_eq!(y.get_a(), 10);
        y.set_a(20);
        assert_eq!(y.get_a(), 20);
        let y2 = Y::add(y, 20);
        assert_eq!(y2.get_a(), 40);

        // test multiple instances
        let y1 = Y::new(1);
        let y2 = Y::new(2);
        assert_eq!(y1.get_a(), 1);
        assert_eq!(y2.get_a(), 2);
        y1.set_a(10);
        y2.set_a(20);
        assert_eq!(y1.get_a(), 10);
        assert_eq!(y2.get_a(), 20);
        let y3 = Y::add(y1, 20);
        let y4 = Y::add(y2, 30);
        assert_eq!(y3.get_a(), 30);
        assert_eq!(y4.get_a(), 50);
        Ok(())
    }
}

impl exports::exports::GuestX for ComponentX {
    fn new(a: i32) -> Self {
        Self {
            val: RefCell::new(a),
        }
    }
    fn get_a(&self) -> i32 {
        *self.val.borrow()
    }
    fn set_a(&self, a: i32) {
        *self.val.borrow_mut() = a;
    }
    fn add(x: X, a: i32) -> X {
        {
            let x = x.get::<ComponentX>();
            x.set_a(x.get_a() + a);
        }
        x
    }
}

static mut NUM_DROPPED_ZS: u32 = 0;

impl exports::exports::GuestZ for ComponentZ {
    fn new(a: i32) -> Self {
        Self { val: a }
    }
    fn get_a(&self) -> i32 {
        self.val
    }

    fn num_dropped() -> u32 {
        unsafe { NUM_DROPPED_ZS + 1 }
    }
}

impl Drop for ComponentZ {
    fn drop(&mut self) {
        unsafe {
            NUM_DROPPED_ZS += 1;
        }
    }
}

impl exports::exports::GuestKebabCase for ComponentKebabCase {
    fn new(a: u32) -> Self {
        Self { val: a }
    }
    fn get_a(&self) -> u32 {
        self.val
    }

    fn take_owned(k: KebabCase) -> u32 {
        k.get::<ComponentKebabCase>().get_a()
    }
}
